### PR TITLE
Bugfix for Sentinel-2 radiance calculation

### PR DESCRIPTION
**Pull request by simonrp84**
This PR fixes the radiance calculation for Sentinel-2 data. Currently, the code uses the method suitable for L1b counts -> radiance computation. However, L1b data is not  (yet) available to users and the procedure is different for L1c data that's available via the Copernicus data ecosystem.

This PR switches to use the correct calculation. 

In addition, I made some changes to the reader code in preparation for the upcoming availability of L1B data for Sentinel-2. This data isn't yet supported by the reader, though, so I have added an explicit error message / reader failure mode if the user tries to process L1B data.

 - [x] Tests added 
 - [x] Fully documented

### CODE REVIEW

File: satpy/readers/msi_safe.py [hunk 1]
```diff
@@ -93,7 +99,21 @@ def _read_from_file(self, key):
...
+                # By default, we use the mean solar angles so that the user does not need to resample,
+                # but the user can also choose to use the solar angles from the tile metadata.
+                # This is on a coarse grid so for most bands must be resampled before use.
+                dq = dict(name="solar_zenith_angle", resolution=key["resolution"])
+                zen = self._tile_mda.get_dataset(dq, dict(xml_tag="Sun_Angles_Grid/Zenith"))
+                tmp_refl = self._mda.calibrate_to_reflectances(proj, self._channel)
+                return self._mda.calibrate_to_radiances(tmp_refl, zen, self._channel)
+            #else:
+                # For L1B the radiances can be directly computed from the digital counts.
+                #return self._mda.calibrate_to_radiances_l1b(proj, self._channel)
```
Comment (mraspaud) [inline satpy/readers/msi_safe.py:114]: Should this be removed?

File: satpy/readers/msi_safe.py [hunk 2]
```diff
@@ -194,14 +214,37 @@ def band_indices(self):
...
+        return self.solar_irradiances[band_index]
+
+    @cached_property
+    def solar_irradiances(self):
+        """Get the TOA solar irradiance values from the metadata."""
+        irrads = self.root.find(".//Solar_Irradiance_List")
+        if irrads is not None:
+            solar_irrad = {int(irr.attrib["bandId"]): float(irr.text) for irr in irrads}
+        else:
+            solar_irrad = {}
```
Comment (mraspaud) [inline satpy/readers/msi_safe.py:237]: Can this be tested?

File: satpy/readers/msi_safe.py [hunk 3]
```diff
@@ -194,14 +214,37 @@ def band_indices(self):
...
+        return solar_irrad
+
+    @cached_property
+    def sun_earth_dist(self):
+        """Get the sun-earth distance from the metadata."""
+        sed = self.root.find(".//U")
+        if sed is not None:
+            return float(sed.text)
+        return -1
```
Comment (mraspaud) [inline satpy/readers/msi_safe.py:246]: I'm not really found of this. Would it be possible to raise an error here and catch it downstream?

**Review by mraspaud — COMMENTED** [review]
Thanks for this, some comments inline.

File: satpy/readers/msi_safe.py (hunk 1 above)
Comment (simonrp84) [reply satpy/readers/msi_safe.py:114]: I'd like to leave that there as a placeholder. Soon the L1b data will be made available to users, and this section of code will then become useful :-)

Comment (mraspaud) [reply satpy/readers/msi_safe.py:114]: Can we uncomment it then?

**Comment by simonrp84** [comment]
@mraspaud I've added a new commit that should address your suggestions / comments above.

**Review by mraspaud — APPROVED** [review]
LGTM

**Pull request merged**